pilot/test/discovery/discovery.go

```diff
@@ -0,0 +1,142 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	proxyconfig "istio.io/api/proxy/v1/config"
+	"istio.io/istio/pilot/adapter/config/memory"
+	"istio.io/istio/pilot/adapter/serviceregistry/aggregate"
+	"istio.io/istio/pilot/model"
+	"istio.io/istio/pilot/platform"
+	"istio.io/istio/pilot/proxy"
+	"istio.io/istio/pilot/proxy/envoy"
+	"istio.io/istio/pilot/test/mock"
+)
+
+var (
+	defaultDiscoveryOptions = envoy.DiscoveryServiceOptions{
+		Port:            8080,
+		EnableProfiling: true,
+		EnableCaching:   true}
+)
+
+func makeMeshConfig() *proxyconfig.MeshConfig {
+	mesh := proxy.DefaultMeshConfig()
+	mesh.MixerAddress = "istio-mixer.istio-system:9091"
+	mesh.RdsRefreshDelay = ptypes.DurationProto(10 * time.Millisecond)
+	return &mesh
+}
+
+// mockController specifies a mock Controller for testing
+type mockController struct{}
+
+func (c *mockController) AppendServiceHandler(f func(*model.Service, model.Event)) error {
+	return nil
+}
+
+func (c *mockController) AppendInstanceHandler(f func(*model.ServiceInstance, model.Event)) error {
+	return nil
+}
+
+func (c *mockController) Run(<-chan struct{}) {}
+
+func buildMockController() *aggregate.Controller {
+	discovery1 := mock.NewDiscovery(
+		map[string]*model.Service{
+			mock.HelloService.Hostname:   mock.HelloService,
+			mock.ExtHTTPService.Hostname: mock.ExtHTTPService,
+		}, 2)
+
+	discovery2 := mock.NewDiscovery(
+		map[string]*model.Service{
+			mock.WorldService.Hostname:    mock.WorldService,
+			mock.ExtHTTPSService.Hostname: mock.ExtHTTPSService,
+		}, 2)
+
+	registry1 := aggregate.Registry{
+		Name:             platform.ServiceRegistry("mockAdapter1"),
+		ServiceDiscovery: discovery1,
+		ServiceAccounts:  discovery1,
+		Controller:       &mockController{},
+	}
+
+	registry2 := aggregate.Registry{
+		Name:             platform.ServiceRegistry("mockAdapter2"),
+		ServiceDiscovery: discovery2,
+		ServiceAccounts:  discovery2,
+		Controller:       &mockController{},
+	}
+
+	ctls := aggregate.NewController()
+	ctls.AddRegistry(registry1)
+	ctls.AddRegistry(registry2)
+
+	return ctls
+}
+
+// PilotConfig gathers the dependencies of the Pilot discovery service into a single configuration structure.
+type PilotConfig struct {
+	ConfigCache      model.ConfigStoreCache
+	ServiceDiscovery model.ServiceDiscovery
+	ServiceAccounts  model.ServiceAccounts
+	Controller       model.Controller
+	Mesh             *proxyconfig.MeshConfig
+}
+
+// WithInMemoryConfigCache updates the config to use an in-memory cache, backed by the given config store.
+func (cfg *PilotConfig) WithInMemoryConfigCache(configStore model.ConfigStore) *PilotConfig {
```

Comment:
Builder-like patterns typically aren't used in go; instead we tend to prefer option APIs. Check out https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis. gRPC's golang implementation also has tons of examples of this style of object configuration/initialization. In particular, the [`Dial` method + `DialOptions`](https://godoc.org/google.golang.org/grpc#Dial) type are a good example.